### ISSUE TITLE
New Rule: `prefer-destructuring`

### ISSUE THREAD

**Issue by 0xCLARITY** [typescript-eslint/typescript-eslint#1931]
I'd like to use the [`prefer-destructuring`](https://eslint.org/docs/rules/prefer-destructuring) rule from ESLint, however, when I try to use array destructuring with an Express request object, I get the following TypeScript Error:

`Type 'ParamsDictionary' must have a '[Symbol.iterator]()' method that returns an iterator.ts(2488)`.

That seems reasonable, but then the base ESLint `prefer-destructuring` rule gives me a warning, even though TypeScript claims I'm not allowed to use array destructuring here.

I think having a type-aware version of this rule would be beneficial.

**Comment by bradzacher** [comment]
could you please give me a self-contained example?

**Comment by hbergren** [comment]
Does this work?

```ts
import { Request, Response, NextFunction } from 'express'

async function example(
  req: Request,
  _res: Response,
  _next: NextFunction,
): Promise<void> {
  // Type 'ParamsDictionary' must have a '[Symbol.iterator]()' method that returns an iterator
  const [param] = req.params

  console.log(param)
}
```

**Comment by bradzacher** [comment]
what was the code _before_ the rule fixed it?
`const param = req.params[0]`?

**Comment by hbergren** [comment]
Yes, which the base ESLint `prefer-destructuring` rule warns against.

**Comment by bradzacher** [comment]
Thanks.
Happy to accept a PR if you'd like to add a type-aware version of the rule to prevent array destructuring of non-iterable types.
In that case, could also implement a fixer, as we know for sure the type is fixable.

**Comment by phaux** [comment]
The correct way to destructure this is
```js
const { 0: param } = req.params
```
because `params` is just an object with numeric properties.

**Comment by danvk** [comment]
Another issue with the eslint `prefer-destructuring` rule in TS is if you apply a type declaration to the resulting variable:

```ts
const obj = {xs: ['a', 'b', 'c']};  // type is {xs: string[]}
const xs: readonly string[] = obj.xs;
xs.push('d');
// ~~~~ Property 'push' does not exist on type 'readonly string[]'.
```

Line two gets flagged by eslint's rule:

![image](https://user-images.githubusercontent.com/98301/90832432-c0fe5a80-e313-11ea-9d19-9f0da3208684.png)

However, applying the quick fix removes the type annotation, an unsafe operation which makes the `push` call pass the type checker:

```ts
const obj = {xs: ['a', 'b', 'c']};  // type is {xs: string[]}
const {xs} = obj;
xs.push('d');  // ok! oh no!
```

A typescript-eslint `prefer-destructuring` rule should ignore assignments with attached type declarations.

**Comment by rafaelss95** [comment]
Isn't it a duplication of #723? This one had a lot more interaction than the former though.

**Comment by bradzacher** [comment]
No - they both want to extend the same extension rule - but they want different things.

THIS issue is about
`const x = (y as {0: unknown)[0]`
being fixed to
`const [x] = (y as {0: unknown})`

#723 is about
`const x: number = (y as {bar: 1}).bar`
being fixed to
`const {x} = (y as {bar: 1})`

**Comment by ilyub** [comment]
Another issue with core eslint-rule:
```ts
export const construct: Function = Reflect.construct;

// Fixed to:
export const { construct } = Reflect;
```
As you can see, core rule auto-fix removes type definition (": Function").

**Comment by seiyab** [comment]
Hi, I'm working on this issue.
For the original problem, I want to clarify the desired behavior.
I'm considering following behavior. Is it reasonable?
```ts
const a = {0: 1};
const b = a[0]; // this is incorrect iff options[0].object is true and enforceForRenamedProperties  is true
const {0: c} = a; // this is always correct


const x = [1];
const y = x[0]; // works same as the current ESLint rule

const v: any = {0: 1};
const w = x[0]; // works same as the current ESLint rule
```

Showing other edge cases are welcome.

**Comment by seiyab** [comment]
I'm also working on https://github.com/typescript-eslint/typescript-eslint/issues/1931#issuecomment-677940850, https://github.com/typescript-eslint/typescript-eslint/issues/1931#issuecomment-1118179511 and https://github.com/typescript-eslint/typescript-eslint/issues/723.
It is almost done in my local repository. Wait for a little while 😉 .

**Comment by JoshuaKGoldberg** [comment]
> I'm considering following behavior. Is it reasonable?

Sorry, I don't follow what you're asking about @seiyab. [There are no rule reports from the base `prefer-destructuring` rule for `const b = a[0];`](https://typescript-eslint.io/play/#ts=5.1.6&fileType=.tsx&code=MYewdgzgLgBAhjAvDA3jADALhgRhgXwG4AoUSWAIyXgG10BdQmAemZigAsBLCGHvsKABOQgKbBYXAGZSYIAA5Qu4CHXoA6EBQBW4ybyhCArqPhgAJnMXLINHBtFgpIIcFEAxFwCVHcALai5gAKQgqiQkqivHwGxqLEQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQA7UQDNFoBaAExQBc4BjK2aASwDsBzdMAbXGxxOgB7aJAA0vPlj7ZIggEYArRPU41YiCdgC%2B46Zk19IiFoWG1EAMWEAlYwEMAtonIAFIbhJUmKVXA16tTQBdXkCtIA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false).

**Comment by seiyab** [comment]
@JoshuaKGoldberg 
I'm sorry for unclearness. (My English might often be broken.)
The base prefere-destructuring  doesn't report for `const b = a[0]` if `array` in options is `false`, as you wrote.
For the following example, I wanted to show the behavior of the **new rule I'm considering** (and have already implemented as a PR).
```ts
const a = {0: 1};
const b = a[0]; // this is incorrect iff options[0].object is true and options[1].enforceForRenamedProperties  is true
```
In this case, the proper destructuring syntax  is `const { 0: b } = a` rather than `const [b] = a` since `a` is not iterable. This should be the primary point in this issue.
The comment following `const b = a[0]` suggests:
- `options[0].object` should control the report rather than `options[1].array`.
- it should report it only if `options[1].enforceForRenamedProperties` because it is renamed from `0` to `b`

Related document I have written: https://github.com/typescript-eslint/typescript-eslint/pull/7117/files#diff-0d683d5803f89de0a159ca7ef7461802afac69ae7fbf459411553aa36e5a7cbaR49-R50

And I asked that whether my thoughts make sense or not.

Does it get clear?

**Comment by seiyab** [comment]
"this will be incorrect iff ..." might be better than "this is incorrect iff ...".

**Comment by tjx666** [comment]
Example bad case for ESLint builtin prefer-destructuring:

![image](https://github.com/typescript-eslint/typescript-eslint/assets/41773861/7a52762e-73b0-4e83-b315-c4ff119698dc)

After destructuring:

![image](https://github.com/typescript-eslint/typescript-eslint/assets/41773861/071d2ca9-6f4b-41e7-8e24-4eaf96d1cbd9)
